File: packages/companion/swift/Sources/BridgethingCompanion/WebappResourceService.swift
import BridgethingGateway
import BridgethingSchema
import Foundation
#if canImport(CryptoKit)
    import CryptoKit
#endif

public enum WebappResourceError: Error, CustomStringConvertible {
    case notStarted
    case staleCacheMissing
    case shaMismatch(expected: String, got: String)
    case cryptoUnavailable
    case domain(WebappError)
    case wire(WireError)

    public var description: String {
        switch self {
        case .notStarted: "webapp resource service not started"
        case .staleCacheMissing: "daemon reported cache current but no cached file exists"
        case let .shaMismatch(expected, got): "resource sha256 mismatch: expected \(expected), got \(got)"
        case .cryptoUnavailable: "CryptoKit unavailable on this platform"
        case let .domain(err): "daemon rejected resource fetch: \(err)"
        case let .wire(err): "resource fetch protocol error: \(err)"
        }
    }
}

public actor WebappResourceService {
    public struct Resolved: Sendable {
        public let url: URL
        public let mime: String?
        public let sha256: String
    }

    private let receiver: TransferReceiver
    private let cacheDir: URL
    private var gateway: BridgethingGateway?

    init(receiver: TransferReceiver, cacheDirectory: URL? = nil) {
        self.receiver = receiver
        let base = cacheDirectory
            ?? FileManager.default.urls(for: .cachesDirectory, in: .userDomainMask).first
            ?? URL(fileURLWithPath: NSTemporaryDirectory())
        cacheDir = base.appendingPathComponent("bridgething-webapp-resources", isDirectory: true)
    }

    func start(gateway: BridgethingGateway) {
        self.gateway = gateway
    }

    public func fetch(deviceId: String, webappId: UUID, kind: WebappResourceKind) async throws -> Resolved {
        guard let gateway else { throw WebappResourceError.notStarted }
        let cached = newestCached(webappId: webappId, kind: kind)
        let result = try await gateway.webapp.resource(
            deviceId: deviceId,
            WebappResource(id: webappId, kind: kind, have: cached?.sha256)
        )
        switch result {
        case let .ok(reply):
            guard let body = reply.body else {
                guard let cached else { throw WebappResourceError.staleCacheMissing }
                let mime = reply.mime ?? Self.mime(forExtension: cached.url.pathExtension)
                return Resolved(url: cached.url, mime: mime, sha256: reply.sha256)
            }
            let data: Data
            switch body {
            case let .inline(bytes):
                data = bytes
            case let .stream(ref):
                await receiver.register(deviceId: deviceId, ref: ref)
                data = try await receiver.collect(ref.id, timeout: .seconds(30))
            }
            if let err = shaError(data, reply.sha256) { throw err }
            let url = try write(webappId: webappId, kind: kind, sha256: reply.sha256, mime: reply.mime, data: data)
            return Resolved(url: url, mime: reply.mime, sha256: reply.sha256)
        case let .domain(err):
            throw WebappResourceError.domain(err)
        case let .protocolError(err):
            throw WebappResourceError.wire(err)
        }
    }

    private func newestCached(webappId: UUID, kind: WebappResourceKind) -> (url: URL, sha256: String)? {
        let prefix = "\(webappId.uuidString)__\(kind.rawValue)__"
        let fm = FileManager.default
        guard let entries = try? fm.contentsOfDirectory(
            at: cacheDir, includingPropertiesForKeys: [.contentModificationDateKey]
        ) else { return nil }
        let matches = entries.filter { $0.lastPathComponent.hasPrefix(prefix) }
        let newest = matches.max { a, b in
            let da = (try? a.resourceValues(forKeys: [.contentModificationDateKey]))?.contentModificationDate ?? .distantPast
            let db = (try? b.resourceValues(forKeys: [.contentModificationDateKey]))?.contentModificationDate ?? .distantPast
            return da < db
        }
        guard let newest else { return nil }
        let stem = newest.deletingPathExtension().lastPathComponent
        guard let sha = stem.components(separatedBy: "__").last, !sha.isEmpty else { return nil }
        return (cacheDir.appendingPathComponent(newest.lastPathComponent), sha)
    }

    private func write(webappId: UUID, kind: WebappResourceKind, sha256: String, mime: String?, data: Data) throws -> URL {
        let fm = FileManager.default
        try fm.createDirectory(at: cacheDir, withIntermediateDirectories: true)
        let name = "\(webappId.uuidString)__\(kind.rawValue)__\(sha256).\(Self.ext(for: mime))"
        let dest = cacheDir.appendingPathComponent(name)
        try data.write(to: dest, options: .atomic)
        let prefix = "\(webappId.uuidString)__\(kind.rawValue)__"
        if let entries = try? fm.contentsOfDirectory(at: cacheDir, includingPropertiesForKeys: nil) {
            for url in entries where url.lastPathComponent.hasPrefix(prefix) && url.lastPathComponent != name {
                try? fm.removeItem(at: url)
            }
        }
        return dest
    }

    private static func ext(for mime: String?) -> String {
        guard let mime = mime?.lowercased() else { return "bin" }
        if mime.contains("svg") { return "svg" }
        if mime.contains("png") { return "png" }
        if mime.contains("jpeg") || mime.contains("jpg") { return "jpeg" }
        if mime.contains("webp") { return "webp" }
        if mime.contains("gif") { return "gif" }
        if mime.contains("html") { return "html" }
        return "bin"
    }

    private static func mime(forExtension ext: String) -> String? {
        switch ext.lowercased() {
        case "svg": "image/svg+xml"
        case "png": "image/png"
        case "jpeg", "jpg": "image/jpeg"
        case "webp": "image/webp"
        case "gif": "image/gif"
        case "html": "text/html"
        default: nil
        }
    }

    private func shaError(_ data: Data, _ expected: String) -> WebappResourceError? {
        #if canImport(CryptoKit)
            let got = SHA256.hash(data: data).map { String(format: "%02x", $0) }.joined()
            guard got.caseInsensitiveCompare(expected) == .orderedSame else {
                return .shaMismatch(expected: expected, got: got)
            }
            return nil
        #else
            return .cryptoUnavailable
        #endif
    }
}

File: packages/companion/kotlin/companion/src/main/kotlin/com/bridgething/companion/WebappResourceService.kt
package com.bridgething.companion

import com.bridgething.gateway.BridgethingGateway
import com.bridgething.gateway.RequestResult
import com.bridgething.gateway.webapp
import com.bridgething.schema.TransferBody
import com.bridgething.schema.WebappResource
import com.bridgething.schema.WebappResourceKind
import java.io.File
import java.io.IOException
import java.nio.file.Files
import java.nio.file.StandardCopyOption
import java.security.MessageDigest
import java.util.UUID
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

public data class WebappResourceFile(
    val file: File,
    val mime: String?,
    val sha256: String,
)

public class WebappResourceService internal constructor(
    private val cacheDir: File,
    private val gateway: BridgethingGateway,
    private val receiver: TransferReceiver,
) {
    private val dir by lazy { File(cacheDir, "bridgething-webapp-resources") }

    public suspend fun fetch(deviceId: String, webappId: UUID, kind: WebappResourceKind): WebappResourceFile? {
        val cached = newestCached(webappId, kind)
        val reply = when (val r = gateway.webapp.resource(deviceId, WebappResource(id = webappId, kind = kind, have = cached?.sha))) {
            is RequestResult.Ok -> r.response
            else -> return null
        }

        val body = reply.body
        if (body == null) {
            val hit = cached ?: return null
            return WebappResourceFile(hit.file, reply.mime ?: mimeForExt(hit.ext), reply.sha256)
        }

        val bytes = when (body) {
            is TransferBody.Inline -> body.data
            is TransferBody.Stream -> receiver.receive(deviceId, body.data)
        }
        val got = sha256Hex(bytes)
        if (got != reply.sha256.lowercase()) {
            throw IOException("webapp $webappId $kind: sha256 $got != reply ${reply.sha256}")
        }

        val dest = File(dir, "${webappId}__${kind.string}__${reply.sha256}.${extForMime(reply.mime)}")
        atomicWrite(dest, bytes)
        pruneOther(webappId, kind, keep = reply.sha256)
        return WebappResourceFile(dest, reply.mime, reply.sha256)
    }

    private data class Cached(val file: File, val sha: String, val ext: String)

    private fun newestCached(webappId: UUID, kind: WebappResourceKind): Cached? {
        val prefix = "${webappId}__${kind.string}__"
        val newest = dir.listFiles { f -> f.isFile && f.name.startsWith(prefix) }
            ?.maxByOrNull { it.lastModified() } ?: return null
        val rest = newest.name.removePrefix(prefix)
        val dot = rest.lastIndexOf('.')
        if (dot <= 0) return null
        return Cached(newest, rest.substring(0, dot), rest.substring(dot + 1))
    }

    private fun pruneOther(webappId: UUID, kind: WebappResourceKind, keep: String) {
        val prefix = "${webappId}__${kind.string}__"
        dir.listFiles { f -> f.isFile && f.name.startsWith(prefix) }?.forEach { f ->
            if (f.name.removePrefix(prefix).substringBefore('.') != keep) f.delete()
        }
    }

    private fun atomicWrite(dest: File, bytes: ByteArray) {
        dir.mkdirs()
        val tmp = File(dir, "${dest.name}.${UUID.randomUUID()}.tmp")
        tmp.writeBytes(bytes)
        runCatching {
            Files.move(tmp.toPath(), dest.toPath(), StandardCopyOption.ATOMIC_MOVE, StandardCopyOption.REPLACE_EXISTING)
        }.onFailure {
            Files.move(tmp.toPath(), dest.toPath(), StandardCopyOption.REPLACE_EXISTING)
        }
    }

    private suspend fun sha256Hex(bytes: ByteArray): String = withContext(Dispatchers.Default) {
        MessageDigest.getInstance("SHA-256").digest(bytes).joinToString("") { "%02x".format(it) }
    }

    private fun extForMime(mime: String?): String =
        mime?.substringBefore(';')?.trim()?.let { EXT_BY_MIME[it] } ?: "bin"

    private fun mimeForExt(ext: String): String? = MIME_BY_EXT[ext]

    private companion object {
        val EXT_BY_MIME = mapOf(
            "image/svg+xml" to "svg",
            "image/png" to "png",
            "image/jpeg" to "jpeg",
            "image/webp" to "webp",
            "image/gif" to "gif",
            "text/html" to "html",
            "text/javascript" to "js",
        )
        val MIME_BY_EXT = EXT_BY_MIME.entries.associate { (mime, ext) -> ext to mime }
    }
}
